src/gil.rs

```diff
@@ -231,9 +271,39 @@ impl GILPool {
             no_send: Unsendable::default(),
         }
     }
+
     pub unsafe fn python(&self) -> Python {
         Python::assume_gil_acquired()
     }
+
+    /// Reduce the lifetime of a reference to be guaranteed to not outlast this GILPool.
+    ///
+    /// This does not change when the original owned reference will be dropped, but can aid in
+    /// reducing problematic lifetimes when working with GILPools manually (see `GILPool::new`).
+    ///
+    /// # Examples:
+    ///
+    /// ```compile_fail
+    /// use pyo3::{Python, GILPool, ObjectProtocol};
+    ///
+    /// let gil = Python::acquire_gil();
+    /// let py = gil.python();
+    /// let obj = py.eval("object()", None, None).unwrap();
+    ///
+    /// let pool2 = unsafe { GILPool::new() };
+    ///
+    /// // This repr would have lifetime tied to `obj` but cannot be safely used after GILPool
+    /// let repr = pool2.reduce_lifetime(obj.repr().unwrap());
+    ///
+    /// drop(pool2);
+    ///
+    /// // Usage of repr is disallowed because pool2 is dead.
+    /// println!("{:?}", repr);
+    /// ```
+    #[inline]
+    pub fn reduce_lifetime<'py, 'a: 'py, T>(&'py self, reference: &'a T) -> &'py T {
```

Comment:
I can understand what this function does, but I don't understand why we expose this function as a public API.
We need `GILPool` when we can't get `Python` without it(and he/she is certain that the caller has GIL).
So this kind of situation does not seem realistic to me :thinking:

Reply:
👍 I will write some docs in `advanced.md` that explain the situations I think there are. Then we can review what API we like for each situation.